### DISCUSSION TITLE
Welcome to microeco question Discussions!

### DISCUSSION THREAD

**Discussion by ChiLiubio** [ChiLiubio/microeco#6] [Q&A]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by shamsalig** [comment]
Hello there can any one help when I use 
dataset$tidy_dataset()
print(dataset)
Now of rows becomes 0??!!

**Reply by ChiLiubio** [reply]
Please check the reply in issues.

**Comment by kardiolipin** [comment]
Hi Chi,

Can I use data produced in SILVAngs to generate microtable object?

**Reply by ChiLiubio** [reply]
Hi, @kardiolipin . Sure. Actually, it is ok to use results from any bioinformatic platform. The most important thing is to make sure the format of the tables same with the example data, such as checking the rownames of OTU/ASV table and sample table.

**Reply by ChiLiubio** [reply]
Remember also cleaning up the taxonomic table using tidy_taxonomy() function. Please see the taxonomic table in tutorial for the unified prefix and format. I have not used the SILVAngs for a long time and can not recall the how the results look like.

**Reply by kardiolipin** [reply]
Thank you!

**Comment by kardiolipin** [comment]
Hi Chi,
Another question.
I am plotting a bar plot with Relative Abundances. And want to show more than 10 taxa. Let say 30 taxa. How can I provide each taxon with personal color? Because, if I simply change  number of ntaxa, I will have many of them (>12) plotted without color.
How can I modify this code?
t1 <- trans_abund$new(dataset = dataset, taxrank = "Family", ntaxa = 30)
t1$plot_bar(others_color = "Grey50",  facet = "Group", xtext_keep = FALSE, legend_text_italic = TRUE)

**Reply by ChiLiubio** [reply]
Hi,
The first parameter in plot_bar is for color assignment. Please see the help document using ?trans_abund.

**Comment by larrouyj** [comment]
Hello,

Thank you for this package it's so nice !

I was wondering how I could :

Extract the cal_group_distance() value with as well the metadata link to the sample?
And do cal_spe_func_perc / plot_spe_func_perc() for each of my treatment ( = Group in the tutorial) and not module as you have in the tutorial.
Thank you !

**Reply by ChiLiubio** [reply]
Hi,

cal_group_distance() return a table that is transformed from the distance matrix. Please see the Return part in the help document of the function.
The default approach in cal_spe_func_perc() is to calculate samples, not module, see the parameter "use_community" in cal_spe_func_perc() function.

**Comment by HCB-M2RG** [comment]
Hello, I have been test driving this package for the last few weeks and submitting my questions as Issues. Thanks for providing this question/answer forum, which seems like a better place for my interactions.

I am working with the trans_nullmodel class. Ultimately, I want to plot and assess betaNTI values in a similar way that your tutorial showed for betaNRI. I have done this before with the picante package but it is nice to have all of these tools pulled together.

I started by working through the example (https://chiliubio.github.io/microeco/#trans_nullmodel_class) but got the same error when attempting to create a trans_beta class (both with betaNRI and betamntd). Here is the code I used/adapted from the tutorial and error message. Any ideas?

> # generate trans_nullmodel object; use 1000 OTUs as example
> t9 <- trans_nullmodel$new(dataset_16S, taxa_number = 1000, add_data = env_data_16S_simple)
> # add betaNRI matrix to beta_diversity list
> dataset_16S$beta_diversity[["betaNRI"]] <- t9$res_ses_betampd
> 
> # create trans_beta class, use measure "betaNRI"
> t10 <- trans_beta$new(dataset = dataset_16S, group = "Environment.short", measure = "betaNRI")
Error in initialize(...) : 
  Input measure should be one of beta_diversity distance in dataset
 




 
--



 
--




 
--

**Reply by ChiLiubio** [reply]
Hi,
After generating t9, 
t9 <- trans_nullmodel$new(dataset_16S, taxa_number = 1000, add_data = env_data_16S_simple)
Do you use cal_ses_betampd  function to calculate betaNRI?
Otherwise, the betaNRI is null, then leading the error, even you assign NULL to beta_diversity.

**Reply by HCB-M2RG** [reply]
that fixed it. I really appreciate how responsive you have been with help!

**Comment by HCB-M2RG** [comment]
Another question for today. I need to get into the ggplot aesthetics to customize some graphs but do not want to go so far as to edit the actual functions provided in the package (however, I imagine I could possibly create a new one if I knew how). Also, it does not look like I can simply edit the code in functions like "plot_alpha".

Here is the example that I am working with:
t3 <- trans_alpha$new(dataset = dataset_16S, group = "Environment.station")
p.alpha.obs <- t3$plot_alpha(add_letter = FALSE, measure = "Observed", shape = "none")

p.alpha.obs

my goal is to remove the shape = group command set inside of your ggboxplot arguments. Otherwise, get too many categories for ggplot to work with and drop data points from my box plot. I tried to simply add shape in as an argument, but that does not work. I can easily add layers, but not strip them away (if that makes sense).

**Reply by ChiLiubio** [reply]
Hi,
Actually, this issue has exist for a long time.
Today, I update the microeco package in github. In the current version, If you do not add shape = "Group" in the plot_alpha(), the plot will not use the shape in aes. The shape parameter can be passed directly to function ggboxplot. Please see the help document and try again.

**Comment by HCB-M2RG** [comment]
Hello,
my team and I have been working with this package for a couple of different data sets. We discovered that if we take the R6 microtable object (called for example dataset_16S) and save it as a new object (e.g., dataset_16S_2), any operation we do to the new object is also performed on the parent data set. This came as a surprise and may have caused some issues in the way we perform analyses when returning to the original dataset_16S. I am sure there is a good reason for this, some words of warning and solutions. Can you help explain?

**Reply by HCB-M2RG** [reply]
we have seen your notes on how to clone an object and will look into this as well. Any feedback on disadvantages or words of warning when cloning the dataset to perform new operations in a separate direction may still be useful.

**Reply by ChiLiubio** [reply]
Hi, as you see, the direct assignment and operation of all R6 object in this package can also change the parent data set, because there are the same object.  It is related with the reference semantics (https://r6.r-lib.org/articles/Introduction.html) in R6. This design is useful in terms of its principles. However, in our data analysis, we are uninvolved in this character generally. So I create the clone function to deep copy the R6 object. Actually, each R6 object have a clone function inside with a deep parameter. They are same. Currently, almost no complicated structure is used in this package development. But this may be useful in the futhure development. For example, Let's imagine that you have an 16S_data1 object and a deep clone 16S_data2. Now, these two object are seperate in terms of all fields. If you find there is a taxonomy error in 16S_data2, now you change 16S_data2. 16S_data1 is not affected, and you should also change 16S_data1 again. If this field is reference semantics, then you only need to change it in 16S_data2.

**Comment by hassantarabai** [comment]
Hello,

I recently started using the package and i admire much its visuallization compared to other tools. I am having trouble setting color values for plot bar function.

I want to use color_palette_20 but even when applying the default settings, i am getting an error  :"  unused argument (color_values = RColorBrewer::brewer.pal(12, "Paired"))". The same applies for replacing RColorBrewer::brewer.pal(12, "Paired") with color_palette_20. However, color_values is working well for heat map function.

t1$plot_bar(
  color_values = RColorBrewer::brewer.pal(12, "Paired"),
  bar_type = "full",
  others_color = "grey90",
  facet = NULL,
  facet2 = NULL,
  order_facet = NULL,
  x_axis_name = NULL,
  order_x = NULL,
  barwidth = NULL,
  use_alluvium = FALSE,
  clustering = FALSE,
  facet_color = "grey95",
  strip_text = 11,
  legend_text_italic = FALSE,
  xtext_type_hor = TRUE,
  xtext_size = 10,
  xtext_keep = TRUE,
  xtitle_keep = TRUE,
  ytitle_size = 17,
  ylab_title = NULL
)
 Error message:
 Error in t1$plot_bar(color_values = RColorBrewer::brewer.pal(12, "Paired"),  : 
  unused argument (color_values = RColorBrewer::brewer.pal(12, "Paired"))

Kindly, any idea what is causing this error and how to deal with it?

Thank you

**Comment by ChiLiubio** [comment]
Hi. Which version are you using?

**Reply by hassantarabai** [reply]
Version 0.11.0

**Comment by ChiLiubio** [comment]
Weird! I do not find out where is the problem.

**Reply by hassantarabai** [reply]
I solved it by fresh installation of microeco and its dependencies

**Comment by hassantarabai** [comment]
Hello,

I am trying to make an alluvium plot for abundance between differet groups and i noticed "others_colors" for after the set number of taxa to show do no work in alluvium. I guess its because there are multiple taxa in others that hinders that, but it might be useful to have others in alluvium and for it to include the sum of abundance of all taxa in "others" per compared group so it can be comapred between different groups. Just a suggestion :)

**Comment by ChiLiubio** [comment]
Hi @hassantarabai ,

Thanks very much for the suggestion. As you described, I have added this choice same with the comon bar plot in parameter bar_type. Now the default alluvium can also show 'others' with bar_type = "full". Changing bar_type can close showing 'others'.

Best
Chi

**Comment by hassantarabai** [comment]
Hello,

I recently updated to microeco version 0.13.0 from 0.11.0 and noticed some error i started getting in the trans_beta analysis when calculating sample distance within groups "cal_group_distance()". The following error is occuring: Error in colnames(res)[2] <- type : replacement has length zero. I used the same input data and commands and it was working in the previouse version. 

Kindly, any idea what is the reason for that? 

Regards,

**Reply by ChiLiubio** [reply]
Hi @hassantarabai 
Could you please send me your data to have a try? I am not sure how this error happened and need to repeat it. This function has a major updating in this version. You can choose to paste the compressed microtable data format here as the tutorial shows (https://chiliubio.github.io/microeco_tutorial/notes.html#save-function) or send me via email (liuchi0426@126.com). Also send me your running steps please. Thanks.

Best,
Chi

**Comment by ChiLiubio** [comment]
Hi @hassantarabai 
Could you please send me your data to have a try? I am not sure how this error happened and need to repeat it. This function has a major updating in this version. You can choose to paste the compressed data format here as the tutorial shows (https://chiliubio.github.io/microeco_tutorial/notes.html#save-function) or send me via email (liuchi0426@126.com). Thanks.

Best,
Chi

**Reply by hassantarabai** [reply]
Hi @ChiLiubio,

Kindly can you provide me with some other email contact if available to send the steps and Rdata rather than [liuchi0426@126.com] as i tried sending it yesterday to the provided email and it gave me the following error:

The email you sent to liuchi0426@126.com was bounced by the system , the following is the relevant information of the bounced email:
--
 
Your message was marked as spam by the system and could not be delivered ........"

Thank you

**Comment by ChiLiubio** [comment]
Hi @hassantarabai 
Could you please directly drag your compressed data here? It is the easiest way. Please have a try.

**Comment by ChiLiubio** [comment]
Hi @hassantarabai 
![图片](https://user-images.githubusercontent.com/20815519/206065630-0570ea54-8325-4810-980b-e9b2b7f7b345.png)

**Reply by hassantarabai** [reply]
@ChiLiubio 

Here it is.

Regards,


[microeco_steps.txt](https://github.com/ChiLiubio/microeco/files/10173479/microeco_steps.txt)
[Hassan_microeco_tocheck.zip](https://github.com/ChiLiubio/microeco/files/10173502/Hassan_microeco_tocheck.zip)

**Comment by ChiLiubio** [comment]
Hi @hassantarabai 

I found the error occurs here 
```
#generating a trans beta object with measure "wei_unifrac"
a0.5 <- trans_beta$new(dataset = att0.5, measure = "wei_unifrac")
#Transform sample distance within group "Species"
a0.5$cal_group_distance()
```
You donot assign the group parameter when creating a0.5, which is necessary for the distance conversion within or between groups.
Another thing is your sample_table in att0.5 only has one column 'SampleID', in which the elements are unique. So it has no means to convert distance as this function is used to compare distances for groups.
Thanks for your issue. I will add a warning for this function when group is not found.

Best,
Chi

**Comment by ChiLiubio** [comment]
Hi,


Maybe merge_samples function is not what you need in this time. This function is used to merge multiple samples of a group to one when the group information is not needed. For example, in venn diagram analysis multiple samples in a group should be viewed as one. After merging samples, the new sample information is the group names, not the raw samples. If you need add more information, you can manually add columns to the sample_table.

I do not find the reply in the Discussion. So I directly write this via my email. It is better to use Issues (https://github.com/ChiLiubio/microeco/issues)  for a new question. Thanks.



Best,

Chi





At 2022-12-07 20:47:21, "Hassan Tarabai" ***@***.***> wrote:

Hello Chi,

Thank you for your prompt response. based on your comments, i pinpointed the error to the step
att0.5 <- clone0.5$merge_samples(use_group = "Species") which is generating a table with one column although the prior step :
"#Create object for convergency analysis
att0.5 <- clone(clone0.5) "
include a table with multiple columns so am not sure why merging of the samples based on species is generating only a table with one column which is why downstream analysis including distance calling is failing.

**Comment by hassantarabai** [comment]
Hello @ChiLiubio 

I would like to suggest an addition of Upset graph based analysis to the microeco package as it can provide clearer visualization for comparison between multiple groups compared to venn analysis.

**Reply by ChiLiubio** [reply]
Hi @hassantarabai 

Do you mean the plot from upset function (https://www.r-bloggers.com/2020/04/upset-plots/)? I designed a simple bar plot like it in the plot_bar function of trans_venn (https://chiliubio.github.io/microeco_tutorial/composition-based-class.html#trans_venn-class), because I think it is enough to only show the bars. I will think about it more.
Thanks very much for your suggestion!

Best,
Chi

**Reply by hassantarabai** [reply]
@ChiLiubio 

Thank you for considering it. Yes its something similar but a more complex Upset graph like this 
![Upset graph with deviation](https://user-images.githubusercontent.com/81158634/219579175-cf9cb53f-df1d-4a0b-9687-6fd7d13d3e1f.png)

which will useful to show the sizes of the compared groups and thier deviation. 
There is an R package (Upestr) that can do something similar http://gehlenborglab.org/research/projects/upsetr/ but i don't think it can do also deviation visualization.

Some web app with multiple output options for Upset visualization include  : http://vcg.github.io/upset/ and https://vdl.sci.utah.edu/upset2/

It would be useful to have it within microeco package as sometimes venn analysis is not good enough for multiple comapred groups.

Regrads

**Comment by bioinfo17** [comment]
Hi,
First of all, congratulations and thank you for generating this excellent package.
I've a major question:
I've 154 unique family in my taxonomic file generated using the command:   
apply(dataset$tax_table, 2, function(x){length(unique(x))})

However, when I plot venn diagram after merging @ family level and grouping samples at a factor level the total number of family in the venn diagram shows 445 and when I looked at the tax table I see that that the total number of family lines correspond to 445 as the taxonomic lines have empty "f__" as I've tidied up the dataset initially by removing undetermined, uninformative etc info from the taxonomy file.

Can you please advise how can we exclude these empty "f__"s while plotting venn diagrams to make meaningful comparisons?
And is there an easy way (txt or excel format) to view what family taxa is shared and unique between groups within a treatment?

Kind regards

**Reply by ChiLiubio** [reply]
It is better to write the detailed issues in the Issues part (https://github.com/ChiLiubio/microeco/issues). Thanks.

**Reply by ChiLiubio** [reply]
Hi. Here is an example with the descriptions.
```
library(microeco)
library(magrittr)
data(dataset)
# clone a new data to avoid changing ogrinial one
tmp <- clone(dataset)
# so do you mean removing all those f__ ?
tmp$tax_table %<>% subset(Family != "f__")
tmp$tidy_dataset()
# merge Family
tmp_merge <- tmp$merge_samples("Group")
tmp_merge <- tmp_merge$merge_taxa("Family")
# then we should use family names to replace OTU/ASV names
rownames(tmp_merge$otu_table) <- rownames(tmp_merge$tax_table) <- tmp_merge$tax_table$Family
# error because there are duplicates in the names, we can add some other info in front of it
rownames(tmp_merge$otu_table) <- rownames(tmp_merge$tax_table) <- paste0(tmp_merge$tax_table$Order, ":", tmp_merge$tax_table$Family)
# then venn analysis
t1 <- trans_venn$new(tmp_merge)
# data_details; save the family taxa to csv file
write.csv(t1$data_details, "Family.csv")
```

**Reply by bioinfo17** [reply]
Thanks very much for the reply - this is exactly what I was after!!

**Comment by Masoleno** [comment]
Hello! Thank you for such a great package :D I have been using it to make network analysis and functional analysis. I was wondering if there is a way to plot relative abundance of taxa associated with a specific functional group? Something similar to t1$plot_func_FR() but instead of plotting the functional redundancy, just a bar chart or heatmap of relative abundance of each functional group? Ideally I would also like to facet the plot by groupings in my sample data, but I think I can do that quite easily using standard ggplot2 syntax right?

**Reply by ChiLiubio** [reply]
Hi. Do you mean the visualiztion like the following one. It is changed based on the online example (https://chiliubio.github.io/microeco_tutorial/other-examples-1.html#faprotax-differential-test).
```
library(microeco)
data(dataset)
t1 <- trans_func$new(dataset)
t1$cal_spe_func(prok_database = "FAPROTAX")
t1$cal_spe_func_perc(abundance_weighted = TRUE)
# it is better to clone a dataset
tmp_mt <- clone(dataset)
# transpose res_spe_func_perc to be a data.frame like taxonomic abundance
tmp <- as.data.frame(t(t1$res_spe_func_perc), check.names = FALSE)
# assign the table back to taxa_abund list for further analysis
tmp_mt$taxa_abund$func <- tmp

# then use other class, select 'func' data
trans_abund$new(dataset = tmp_mt, taxrank = "func", ntaxa = 10)$
    plot_bar(others_color = "grey70", facet = c("Group"), bar_full = FALSE)
```
Another gentle reminder is for new questions, it is recommended to use Issues (https://github.com/ChiLiubio/microeco/issues) or a new Discussion. This way, we can locate the topic more easily and avoid disturbing others (they may receive emails about this issue) following this thread. Thanks.
